### PR TITLE
fix behavior of ConcatDataset w/ negative indices

### PR DESCRIPTION
**Pull request by jayleverett**
Currently, when you pass a negative index to a `Dataset` created with `ConcatDataset`, it simply passes that index to the first dataset in the list. So if, for example, we took `concatenated_dataset[-1]`, this will give us the last entry of the *first* dataset, rather than the last entry of the *last* dataset, as we would expect.

This is a simple fix to support the expected behavior for negative indices.

### CODE REVIEW

File: torch/utils/data/dataset.py [hunk 1]
```diff
@@ -73,6 +73,8 @@ def __len__(self):
         return self.cumulative_sizes[-1]
 
     def __getitem__(self, idx):
+        if idx < 0:
+            idx = self.__len__() + idx
```
Comment (apaszke) [inline torch/utils/data/dataset.py]: Please use `len(self)`. Also, it would be nice to add a check for `-idx <= len(self)` and raise an error otherwise

**Review by apaszke — APPROVED** [review]

**Comment by fmassa** [comment]
@jayleverett could you address Adam's request?

File: torch/utils/data/dataset.py (hunk 1 above)
Comment (jayleverett) [reply torch/utils/data/dataset.py]: Hi Adam - sorry I missed this. Added another commit.

File: torch/utils/data/dataset.py [hunk 2]
```diff
@@ -73,6 +73,9 @@ def __len__(self):
         return self.cumulative_sizes[-1]
 
     def __getitem__(self, idx):
+        if idx < 0:
+            assert -idx <= len(self), "absolute value of index should not exceed dataset length"
```
Comment (apaszke) [inline torch/utils/data/dataset.py]: It would be slightly better to raise `ValueError` instead of an `AssertionError` here.

**Review by apaszke — APPROVED** [review]

**Comment by fmassa** [comment]
Hi @jayleverett , could you address Adam's comment? After that and some tests, this is good to merge. Thanks!

File: torch/utils/data/dataset.py (hunk 2 above)
Comment (jayleverett) [reply torch/utils/data/dataset.py]: Done.

File: torch/utils/data/dataset.py [hunk 3]
```diff
@@ -73,6 +73,10 @@ def __len__(self):
         return self.cumulative_sizes[-1]
 
     def __getitem__(self, idx):
+        if idx < 0:
+            if -idx > len(self):
+             raise ValueError("absolute value of index should not exceed dataset length")
```
Comment (fmassa) [inline torch/utils/data/dataset.py]: linter is failing here (indentation is not a multiple of 4).
Could you look into fixing it?

Comment (jayleverett) [reply torch/utils/data/dataset.py]: Oh weird - sorry about that. Fixed.

**Pull request closed without merging**